File: frontend/src/features/webtoon-cut/readingOrder.ts
import type { PixelRegion } from "./types";

export function orderReadingSequence<T extends PixelRegion>(regions: readonly T[]): T[] {
  const sorted = [...regions].sort((a, b) => a.y0 - b.y0 || a.x0 - b.x0 || a.y1 - b.y1 || a.x1 - b.x1);
  const rows: T[][] = [];

  for (const region of sorted) {
    const row = rows.find((candidate) => sameVisualRow(region, candidate));
    if (row) row.push(region);
    else rows.push([region]);
  }

  return rows
    .sort((a, b) => Math.min(...a.map((item) => item.y0)) - Math.min(...b.map((item) => item.y0)))
    .flatMap((row) => row.sort((a, b) => a.x0 - b.x0 || a.y0 - b.y0 || a.x1 - b.x1 || a.y1 - b.y1));
}

function sameVisualRow(region: PixelRegion, row: readonly PixelRegion[]) {
  const rowY0 = Math.min(...row.map((item) => item.y0));
  const rowY1 = Math.max(...row.map((item) => item.y1));
  const overlap = Math.max(0, Math.min(region.y1, rowY1) - Math.max(region.y0, rowY0));
  const regionHeight = Math.max(1, region.y1 - region.y0);
  const rowHeight = Math.max(1, rowY1 - rowY0);
  const centerDelta = Math.abs((region.y0 + region.y1) / 2 - (rowY0 + rowY1) / 2);
  const tolerance = Math.min(regionHeight, rowHeight);
  return overlap >= tolerance * 0.35 || centerDelta < tolerance * 0.5;
}

File: backend/app/services/webtoon_panel_engine/reading_order.py
from __future__ import annotations

from typing import Iterable, TypeAlias

Box: TypeAlias = tuple[int, int, int, int]


def order_reading_sequence(boxes: Iterable[Box]) -> list[Box]:
    """Return boxes in visual reading order: top-to-bottom, then left-to-right."""
    sorted_boxes = sorted(list(boxes), key=lambda box: (box[1], box[0], box[3], box[2]))
    rows: list[list[Box]] = []

    for box in sorted_boxes:
        for row in rows:
            if _same_visual_row(box, row):
                row.append(box)
                break
        else:
            rows.append([box])

    ordered: list[Box] = []
    for row in sorted(rows, key=lambda items: min(item[1] for item in items)):
        ordered.extend(sorted(row, key=lambda item: (item[0], item[1], item[2], item[3])))
    return ordered


def _same_visual_row(box: Box, row: list[Box]) -> bool:
    row_y0 = min(item[1] for item in row)
    row_y1 = max(item[3] for item in row)
    box_y0, box_y1 = box[1], box[3]
    overlap = max(0, min(box_y1, row_y1) - max(box_y0, row_y0))
    box_height = max(1, box_y1 - box_y0)
    row_height = max(1, row_y1 - row_y0)
    center_delta = abs(((box_y0 + box_y1) / 2) - ((row_y0 + row_y1) / 2))
    row_tolerance = min(box_height, row_height)
    return overlap >= row_tolerance * 0.35 or center_delta < row_tolerance * 0.5
